pkg/controller/compute/v1beta1/cluster.go

```diff
@@ -0,0 +1,540 @@
+		cr.Status.SetConditions(v1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1beta1.ClusterStateProvisioning:
+		cr.Status.SetConditions(v1alpha1.Creating())
+	case v1beta1.ClusterStateUnspecified, v1beta1.ClusterStateDegraded, v1beta1.ClusterStateError, v1beta1.ClusterStateReconciling:
+		cr.Status.SetConditions(v1alpha1.Unavailable())
+	}
+
+	u, _ := gke.IsUpToDate(&cr.Spec.ForProvider, *existing)
+
+	return resource.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  u,
+		ConnectionDetails: connectionDetails(existing),
+	}, nil
+}
+
+func (e *clusterExternal) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	cr, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errNotCluster)
+	}
+	cr.SetConditions(v1alpha1.Creating())
+
+	// Generate GKE cluster from resource spec.
+	cluster := gke.GenerateCluster(cr.Spec.ForProvider)
+
+	// Insert default node pool for bootstrapping cluster. This is required to
+	// create a GKE cluster. After successful creation we delete the bootstrap
+	// node pool immediately and provision any subsequent node pools using the
+	// NodePool resource type.
+	gke.GenerateNodePoolForCreate(cluster)
+
+	create := &container.CreateClusterRequest{
+		Cluster: cluster,
+	}
+
+	if _, err := e.cluster.Projects.Locations.Clusters.Create(gke.GetFullyQualifiedParent(e.projectID, cr.Spec.ForProvider), create).Context(ctx).Do(); err != nil {
+		return resource.ExternalCreation{}, errors.Wrap(err, errCreateCluster)
+	}
+
+	return resource.ExternalCreation{}, nil
+}
+
+func (e *clusterExternal) Update(ctx context.Context, mg resource.Managed) (resource.ExternalUpdate, error) {
+	cr, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalUpdate{}, errors.New(errNotCluster)
+	}
+
+	// We have to get the cluster again here to determine how to update.
+	existing, err := e.cluster.Projects.Locations.Clusters.Get(gke.GetFullyQualifiedName(e.projectID, cr.Spec.ForProvider)).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalUpdate{}, errors.Wrap(err, errGetCluster)
+	}
+
+	u, kind := gke.IsUpToDate(&cr.Spec.ForProvider, *existing)
+	if u {
+		return resource.ExternalUpdate{}, nil
+	}
+
+	_, err = updateFactory(kind, &cr.Spec.ForProvider)(ctx, e.cluster, gke.GetFullyQualifiedName(e.projectID, cr.Spec.ForProvider))
+	return resource.ExternalUpdate{}, errors.Wrap(err, errUpdateCluster)
+}
+
+func (e *clusterExternal) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return errors.New(errNotCluster)
+	}
+	cr.SetConditions(runtimev1alpha1.Deleting())
+
+	_, err := e.cluster.Projects.Locations.Clusters.Delete(gke.GetFullyQualifiedName(e.projectID, cr.Spec.ForProvider)).Context(ctx).Do()
+	return errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errDeleteCluster)
+}
+
+// connectionSecret return secret object for cluster instance
```

Comment:
In general I agree with @muvaf about putting more glue code in the client package, but I'm okay with the current implementation here.

If you think about the responsibilities each package has, frequently it's that the client package forms the data used to CRUD the external resource, while the controller package executes the actual update call. In some cases both responsibilities are in the client package, which presents its own client interface for the controller to call. Typically only one package knows anything about the actual cloud provider SDK calls, which is a good separation of responsibility between packages.

Initially in this PR the client package was forming the CRUD data, and the controller was executing the update calls, but due to the complex way in which GKE updates work the details of how GKE SDK was called were leaked into the client package. We solved this by moving that detail into the controller package - another (possibly better) option would have been to move _everything_ into the client package and have it present a CRUD API to the controller package, which would then not need to be aware of how the GKE SDK works. That was likely my fault, because my comment opened with "we should move everything into the controller package" before clarifying the actual intent.

So it _might_ be a little better to move everything into the client package, but I don't think it's worth the refactoring effort at this stage.